homeassistant/components/mqtt/notify.py

```diff
@@ -0,0 +1,393 @@
+        _LOGGER.error(
+            "Notify service '%s' already exists, cannot register service(s)",
+            service_name,
+        )
+        async_dispatcher_send(hass, MQTT_DISCOVERY_DONE.format(discovery_hash), None)
+        clear_discovery_hash(hass, discovery_hash)
+        return
+    device_id = None
+    if CONF_DEVICE in config:
+        await _update_device(hass, config_entry, config)
+
+        device_registry = await hass.helpers.device_registry.async_get_registry()
+        device = device_registry.async_get_device(
+            {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
+            {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
+        )
+
+        device_id = device.id
+
+    service = MqttNotificationService(
+        hass,
+        config[CONF_COMMAND_TOPIC],
+        MqttCommandTemplate(config.get(CONF_COMMAND_TEMPLATE), hass=hass),
+        config[CONF_ENCODING],
+        config.get(CONF_NAME),
+        config[CONF_QOS],
+        config[CONF_RETAIN],
+        config[CONF_TARGETS],
+        config[CONF_TITLE],
+        discovery_hash=discovery_hash,
+        device_id=device_id,
+        config_entry=config_entry,
+    )
+    services[service_name] = service
+
+    await service.async_setup(hass, service_name, service_name)
+    await service.async_register_services()
+
+
+def has_notify_services(hass: HomeAssistant, device_id: str) -> bool:
+    """Check if the device has registered notify services."""
+    services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+    for key, service in services.items():  # pylint: disable=unused-variable
+        if service.device_id == device_id:
+            return True
+    return False
+
+
+async def async_get_service(
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> MqttNotificationService | None:
+    """Prepare the MQTT notification service through configuration.yaml."""
+    name = config.get(CONF_NAME)
+    if CONF_NAME not in config:
+        config[CONF_NAME] = DOMAIN
+    service_name = slugify(name or DOMAIN)
+    has_services = hass.services.has_service(notify.DOMAIN, service_name)
+    services = hass.data.setdefault(MQTT_NOTIFY_SERVICES_SETUP, {})
+    if service_name in services.keys() or has_services:
+        _LOGGER.error(
+            "Notify service '%s' is not unique, cannot register service(s)",
+            service_name,
+        )
+        return None
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    services[service_name] = MqttNotificationService(
+        hass,
+        config[CONF_COMMAND_TOPIC],
+        MqttCommandTemplate(config.get(CONF_COMMAND_TEMPLATE), hass=hass),
+        config[CONF_ENCODING],
+        name,
+        config[CONF_QOS],
+        config[CONF_RETAIN],
+        config[CONF_TARGETS],
+        config[CONF_TITLE],
+    )
+    return services[service_name]
+
+
+class MqttNotificationServiceUpdater:
```

Comment:
It's not ideal that device_trigger, scene, tag, and now notify implement their own versions of discovery update.
We should move that to a separate helper class, similar to what we have for entities.

Reply:
The cause is here that there is no entity involved. May be it is smart to move this part to a different PR.

Reply:
We plan this in a separate PR